Sv2 ipc

Sjors opened the pull request:
Based on https://github.com/bitcoin/bitcoin/pull/30475 and https://github.com/bitcoin/bitcoin/pull/30437.

Alternative to https://github.com/bitcoin/bitcoin/pull/29432

Usage:

Start node built with `/configure --enable-multiprocess`:

```
src/bitcoin-node -chain=main -ipcbind=unix -listen=0
```

Start the Template Provider:

```
src/bitcoin-mine -debug=sv2 -loglevel=sv2:trace
```

See general Stratum v2 usage instructions in #30437.

To test "real" mining it may be helpful to cherry-pick testnet4 commits.

Sjors commented:
@ryanofsky I slightly mangled e3736422b87e9f5b67d058fc278fe0838e073e07 because of https://github.com/bitcoin/bitcoin/pull/30437/commits/4e1a4342f3b2a857e3211587cd14e36704197483#r1681193702. I'll clean that up later.

Sjors commented:
Incorporated the changes from https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2233717971 and cleaned up commit history.

Sjors commented:
@ryanofsky do you have a commit somewhere that adds libmultiprocess to the guix build?

ryanofsky commented:
re: https://github.com/Sjors/bitcoin/pull/48#issuecomment-2235989334

> @ryanofsky do you have a commit somewhere that adds libmultiprocess to the guix build?

I don't, since I haven't really experimented with the guix build. Maybe simplest approach to start with would enable the multiprocess build option and build and package all the multiprocess executables including bitcoin-tp and bitcoin-node. I know in the longer run you want to have separate releases as described in https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2226847354.

re: https://github.com/Sjors/bitcoin/pull/48#issuecomment-2236061272

> Getting a cryptic `error: '__NVCC___WORKAROUND_GUARD' is not defined` error for the multiprocess job: https://cirrus-ci.com/task/4940413969104896?logs=ci#L2225
> 
> Somehow related to the Template Provider depending on `boost/multi_index` via its use of the mempool.

This is a `-Wundef` error caused by https://github.com/bitcoin/bitcoin/pull/29876 that happens when a boost header is included from a source file that is compiled with `-I /path/to/boost/headers` rather than `-isystem /path/to/boost/headers`. The fix for this is probably:

```diff
-bitcoin_mine_CPPFLAGS = $(bitcoin_bin_cppflags)
+bitcoin_mine_CPPFLAGS = $(bitcoin_bin_cppflags) $(BOOST_CPPFLAGS)
```
 
to add the right -isystem option.

> I can't make sense of the `personality(old_personality | ADDR_NO_RANDOMIZE)` error thrown by MSAN and TSAN, but it's possible that's related to my self-hosted CI setup and/or the modification in [3319668](https://github.com/Sjors/bitcoin/commit/33196682f15caf6fb54d05ed4c53a0d711a91241).

Not sure about this error. Link seems to be https://cirrus-ci.com/task/4759864851824640?logs=ci#L2486

Sjors commented:
The MSan error is introduced by the TemplateProvider in #49. It's not present in Sv2Connman, see #50. Although this runner is flaky on my machine and occasionally needs a restart, this failure seems fairly consistent.

Sjors commented:
I'm getting the MSAN / TSAN error inconsistently on my self-hosted CI. Does not seem related to this PR.

Sjors commented:
Rebased on the latest prerequisite PRs. Not thoroughly tested.

Sjors commented:
I dropped some changes that explicitly disable windows. That said it's untested, and the multiprocess dependency can't be built atm: `MULTIPROCESS=1 HOSTS='x86_64-w64-mingw32' ./contrib/guix/guix-build` will complain it can't `#include <sys/resource.h>`.

Sjors commented:
Rebased for testing https://github.com/bitcoin/bitcoin/pull/30510.

The change to cmake invalidated a bunch of changes in 3cc9640. For now I just dropped them, but I'll have to bring some back later.

It's a bit of a mess now. E.g. building fails with:

```
/Users/sjors/dev/bitcoin-sv2-ipc/src/txmempool.h:26:10: fatal error: 'boost/multi_index/hashed_index.hpp' file not found
#include <boost/multi_index/hashed_index.hpp>
```

ryanofsky commented:
> Which is something I previously fixed for autotools: [#48 (comment)](https://github.com/Sjors/bitcoin/pull/48#issuecomment-2237477471)

You may be able to fix that by adding `Boost::headers` to `target_link_libraries(bitcoin-mine ...)`

Sjors commented:
@ryanofsky I ended up moving `#include <node/miner.h>` to the tests, which is the only place that needs it. Changed in #49.

I also moved `TemplateProvider` from node to common to avoid having to pull in `bitcoin_node`. It could probably be its own lib, but I'd like to see something like #47 happen first.

Tested locally by connecting SRI and mining a testnet4 block.

I noticed on macOS 14.6.1 that `bitcoin-node -testnet4 -ipcbind=unix` listens on `unix:/Users/sjors/Library/Application Support/Bitcoin/testnet4/node.sock`, as expected. But `bitcoin-mine -testnet4` (or `-chain=testnet4`) won't connect unless I give it that path. It's also ignoring the default sv2 listening port for testnet4. So there's probably something wrong with chainparam network selection

Sjors commented:
I set `-DENABLE_WALLET=OFF` for more CI machines.

I also added `or WITH_MULTIPROCESS` to `find_package(Libevent`, though not sure if that's needed.

Guix depends (cmake) build currently fails:

```
make: Entering directory '/bitcoin/depends'
copying packages: native_libmultiprocess native_capnp boost libevent systemtap libmultiprocess capnp
to: /bitcoin/depends/x86_64-linux-gnu
make: Leaving directory '/bitcoin/depends'
-- The CXX compiler identification is GNU 12.4.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /home/guix/.guix-profile/bin/x86_64-linux-gnu-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at /bitcoin/depends/x86_64-linux-gnu/native/lib/cmake/Libmultiprocess/LibmultiprocessConfig.cmake:46 (include):
  include could not find requested file:

    /bitcoin/depends/x86_64-linux-gnu/native/lib/cmake/Libmultiprocess/LibTargets.cmake
Call Stack (most recent call first):
  CMakeLists.txt:172 (find_package)
```

This PR stack is a bit behind master, so maybe it's already fixed. cc @ryanofsky does this look familiar?

ryanofsky commented:
> /bitcoin/depends/x86_64-linux-gnu/native/lib/cmake/Libmultiprocess/LibTargets.cmake

I'll take a look at your branch but this error stands out because this this file is not expected to exist. When cross compiling, libmultiprocess library targets `LibTargets` should not be built, installed or needed in the native build, only in the cross compiled build. That's why bitcoin/CMakeLists.txt writes:

```cmake
  find_package(Libmultiprocess COMPONENTS Lib)
  find_package(LibmultiprocessNative COMPONENTS Bin NAMES Libmultiprocess)
```

And only looks for the "Bin" component not the "Lib" component in the native build. If the native find_package line did include "Lib" you would expect to see the error above.

ryanofsky commented:
I guess looking at CMakeLists.txt in the sv2-ipc branch I don't see any changes to the find_package calls that could cause this. And I also don't think it's possible that 9378b30774d40c8a093c69d2b1807c9eea32455a could affect that find_package call at all, but maybe changing the BUILD_TESTS options could. I wonder if you were able to find a workaround for the problem, or if not if you are still saying the same native LibTargets error currently.

Sjors commented:
Since the problem went away once I rebased on master, and applied the `BUILD_TESTS` fix, I haven't looked for another workaround. I'll just try again after the next rebase round.

Sjors commented:
> Since the problem went away once I rebased on master, and applied the `BUILD_TESTS` fix, I haven't looked for another workaround. I'll just try again after the next rebase round.

I probably did something wrong, because the issue is still there after a full rebase.

It's unrelated though: I tracked it down to the introduction of cmake, see #30931.

Sjors commented:
Rebased after #30510.

Sjors commented:
Still a few things left to clean up, but this now produces a working `bitcoin-mine` for `x86_64-darwin`.

Sjors commented:
Rebased. The guix build for `bitcoin-mine` build now works for all non-Windows platforms. On Apple Silicon machines it needs `codesign -s - bitcoin-mine`.

Some CI failures are expected due to https://github.com/Sjors/bitcoin/pull/49#issuecomment-2493825685

Sjors commented:
Here's a release binary to try it: https://github.com/Sjors/bitcoin/releases/tag/sv2-tp-0.1.10-ipc

Sjors commented:
New release tag, binaries underway: https://github.com/Sjors/bitcoin/releases/tag/sv2-tp-0.1.12-ipc

Sjors commented:
New release tag: https://github.com/Sjors/bitcoin/releases/tag/sv2-tp-0.1.13-ipc

Sjors commented:
Meanwhile in Bitcoin Core, it looks like `libmultiprocess` will be included as a git subtree, see https://github.com/bitcoin/bitcoin/pull/31741. I'll soon give this PR a rebase, using https://github.com/bitcoin/bitcoin/pull/31802 and the latest changes in #49. Once that works, I'll tag a new release and generate fresh `bitcoin-node` and `bitcoin-mine` binaries.

Sjors commented:
New IPC release: https://github.com/Sjors/bitcoin/releases/tag/sv2-tp-0.1.14-ipc

Sjors commented:
Rebased. Now that https://github.com/bitcoin/bitcoin/pull/31741 and some other things landed, this PR is a lot simpler.

New release: https://github.com/Sjors/bitcoin/releases/tag/sv2-tp-0.1.17-ipc

Sjors commented:
Note to self: the fixup commit needs to be squashed into 82e1a80c58a6b6217911654f995d8380adf333c0.

Sjors commented:
https://github.com/bitcoin/bitcoin/pull/31802 landed!

Last rebase before I'm going to split off the sv2 functionality into a fresh codebase.
